homeassistant/components/solarman/config_flow.py

```diff
@@ -0,0 +1,221 @@
+from homeassistant.helpers.service_info.zeroconf import ZeroconfServiceInfo
+
+from .const import (
+    CONF_FW,
+    CONF_FW_VERSION,
+    CONF_PRODUCT_TYPE,
+    CONF_SERIAL,
+    CONF_SN,
+    DEFAULT_PORT,
+    DOMAIN,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class SolarmanConfigFlow(ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Solarman."""
+
+    VERSION = 1
+
+    host = ""
+    model = ""
+    device_sn = ""
+    fw_version = ""
+
+    def __init__(self) -> None:
+        """Initialize the config flow."""
+        self.client: Solarman
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle the initial step via user interface."""
+        errors = {}
+        if user_input is not None:
+            self.host = user_input[CONF_HOST]
+            port = user_input.get(CONF_PORT, DEFAULT_PORT)
+
+            self.client = Solarman(async_get_clientsession(self.hass), self.host, port)
+
+            try:
+                config_data = await self.client.get_config()
+
+                device_info = config_data.get(CONF_DEVICE, config_data)
+
+                self.device_sn = device_info.get(CONF_SN, None)
+                self.fw_version = device_info.get(CONF_FW, None)
+                self.model = device_info.get(CONF_TYPE, None)
+
+            except TimeoutError:
+                errors["base"] = "timeout"
+            except ConnectionError:
+                errors["base"] = "cannot_connect"
+            except Exception:
+                _LOGGER.exception("Unknown error occurred while verifying device")
+                errors["base"] = "unknown"
+            else:
+                await self.async_set_unique_id(self.device_sn)
+                self._abort_if_unique_id_configured()
+
+                if not errors:
+                    return self.async_create_entry(
+                        title=f"{self.model} ({self.host})",
+                        data={
+                            CONF_HOST: self.host,
+                            CONF_PORT: port,
+                            CONF_SN: self.device_sn,
+                            CONF_FW_VERSION: self.fw_version,
+                            CONF_MODEL: self.model,
+                        },
+                    )
+
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(CONF_HOST): str,
+                    vol.Optional(CONF_PORT, default=DEFAULT_PORT): int,
+                }
+            ),
+            errors=errors,
+        )
+
+    async def async_step_reconfigure(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle reconfiguration of the integration."""
+        errors: dict[str, str] = {}
+        reconfigure_entry = self._get_reconfigure_entry()
+
```

Comment:
Let's cut out reconfigure for now as it will speed up the review, we can add this in a later PR